File: day_06/day6.jl
#=
day6:
- Julia version: 
- Author: mike.downey
- Date: 2022-12-06
=#
 
data = read(open("day_06/data.txt"))

function aretheyunique(letters)
    s = Set([l for l in letters])
    return length(s) == length(letters)
end


function get_marker(word, len)
    for i in range(1, length(word))
        if aretheyunique(word[i : i + len - 1]) # remember index starts at 1 so
            return i + len - 1 # skip 'len' and report first char after 'start' marker
        end
    end
end

# Part 1
println("Part 1")
println(get_marker(data, 4))

# Part 2
println("Part 2")
println(get_marker(data, 14))

File: day_06/day6.py
from aoc import aoc


def aretheyunique(letters):
    s = set([l for l in letters])
    return len(s) == len(letters)


def get_marker(word, length=4):
    for i in range(len(word)):
        if aretheyunique(word[i:i + length]):
            return i + length  # skip 4 and report first char after 'start' marker


def part1(data):
    return get_marker(data)


def part2(data):
    return get_marker(data, 14)


def run(file):
    print("Part 1")
    data = aoc.load(file)
    print(part1(data))
    print("Part 2")
    print(part2(data))


if __name__ == "__main__":
    print("Test Data")
    run("test.txt")
    print("Actual Data")
    run("data.txt")
